xla/service/gpu/tree_reduction_rewriter.cc

```diff
@@ -86,6 +86,29 @@ class ReductionRewriterVisitor : public DfsHloRewriteVisitor {
     return false;
   }
 
+  // We observe larger n_div_k can improve tree reduction performance. Swap k
+  // and n_div_k if possible.
+  bool ShouldSwapInnerAndOuterReducedMinorDimension(uint64_t k,
+                                                    uint64_t n_div_k,
+                                                    uint64_t n,
+                                                    int64_t race_free_bound,
+                                                    bool is_row_reduction) {
+    CHECK(k >= n_div_k);
+    // Keep inner reduction as race free.
+    if (k > race_free_bound) {
+      return false;
+    }
+    // Rough conditions for row reduction vectorization, not mean that
+    // vectorization will definitely occur.
+    bool maybe_vectorized = n_div_k % 2 == 0 && n % 2 == 0;
+    // Swapping only affects row reduction vectorization.
+    if (is_row_reduction && maybe_vectorized) {
+      // Swap if n_div_k is small enough or k is divisible by 2 also.
+      return n_div_k * 2 < k || k % 2 == 0;
+    }
+    return true;
```

Comment:
Why do we always swap in this case?

Reply:
Inner reduction has same amounts of load whether swap or not. In this case, vectorization can not enable and k is larger than n_div_k. After swapping, load spend same time, but the store can be reduced. And k is related about how many blocks will be launched. For example, assume row reduction input shape is [b, k, n_div_k], the launched blocks are O(b * k). After swapping, block launch overhead can be reduced also.